### DISCUSSION TITLE
I have a problem with Coqui TTS API

### DISCUSSION THREAD

**Discussion by AliA6565** [coqui-ai/TTS#4008] [General Q&A] [answered]
Hello, I have a porblem with Coqui TTS API, I tried to put it to my chatbot, but every time it's generating a voice, I get this error: Error with Coqui TTS: Expected tuple of length 2. Received: None

**Comment by eginhard** [comment, marked as answer]
What's the command you're trying to run and what's the full error message?

**Reply by AliA6565** [reply]
import os
import time
import pygame
import speech_recognition as sr
from gtts import gTTS
from dotenv import load_dotenv
from gradio_client import Client

# Initialize Pygame mixer for audio playback
pygame.mixer.init()

# Load the environment variables for RVC
load_dotenv()

# Load the LLM (Qwen API) for chat responses
qwen_client = Client("https://qwen-qwen2-5-72b-instruct.hf.space/")

# Function to select microphone
def select_microphone():
    r = sr.Recognizer()
    mic_list = sr.Microphone.list_microphone_names()

    print("Available microphones:")
    for i, mic_name in enumerate(mic_list):
        print(f"{i}: {mic_name}")

    mic_index = int(input("Enter the number of the microphone you want to use: "))
    return mic_index

# Function to transcribe speech
def transcribe_speech(mic_index):
    r = sr.Recognizer()
    with sr.Microphone(device_index=mic_index) as source:
        r.adjust_for_ambient_noise(source, duration=1)
        print("Listening... please speak.")
        audio = r.listen(source)

    try:
        print("Transcribing audio...")
        return r.recognize_google(audio)
    except sr.UnknownValueError:
        return "Sorry, I couldn't understand."
    except sr.RequestError as e:
        return f"Error with the service; {e}"

# Function to generate response from the Qwen API
def generate_response(user_input):
    result = qwen_client.predict(
        query=user_input,
        history=[],
        system="You are Hatsune Miku, a friendly virtual assistant who loves music and chatting with friends. Please reply sweetly and gently!",
        api_name="/model_chat"
    )
    return result[1][0][1]

# Function to use GTTS for text-to-speech and run through Coqui TTS
def coqui_tts(text, reference_audio_url):
    coqui_client = Client("https://coqui-xtts.hf.space/--replicas/mvdvr/")
    print(f"Sending text to Coqui TTS API: {text}")
    print(f"Using reference audio URL: {reference_audio_url}")
    try:
        result = coqui_client.predict(
            text,  # Text to synthesize
            "en,en",  # Language
            reference_audio_url,  # Reference audio URL
            reference_audio_url,  # Reference mic audio (using the same URL here)
            True,  # Use microphone
            True,  # Cleanup reference voice
            True,  # Disable language auto-detect
            True,  # Agree to terms
            fn_index=1
        )
        print("TTS result:", result)
    except Exception as e:
        print(f"Error with Coqui TTS: {e}")

# Main chatbot loop
def chatbot():
    mic_index = select_microphone()
    reference_audio_url = "https://raw.githubusercontent.com/AliA6565/audio/main/1.wav"
    
    print("Welcome! I'm Hatsune Miku, your virtual friend. Let's chat! 😊")

    while True:
        user_input = transcribe_speech(mic_index).strip()
        print(f"You said: {user_input}")

        if user_input.lower() in ['exit', 'bye', 'goodbye']:
            print("Hatsune Miku: Goodbye! It was nice chatting with you! 🎶")
            break

        # Generate response from Qwen API
        ai_response = generate_response(user_input)
        print(f"Hatsune Miku is responding: {ai_response}")

        # Convert text to speech using Coqui TTS with voice cloning from GitHub-hosted reference audio
        coqui_tts(ai_response, reference_audio_url)

if __name__ == "__main__":
    chatbot()

**Reply by eginhard** [reply]
Ok, so you're not actually using this library, so I can't help you. If you want to use Coqui TTS from Python, see the docs: https://coqui-tts.readthedocs.io/en/latest/inference.html#python-tts-api

**Reply by AliA6565** [reply]
thank you

**Comment by AliA6565** [comment]
Thanks
